File: ABCDE.py
# 백준 13023번 ABCDE

n, m = map(int,input().split())

friends = [[]*i for i in range(n)]

for _ in range(m):
    f1, f2 = map(int,input().split())
    friends[f1].append(f2)
    friends[f2].append(f1)

# 방문됐는지 
visited = [False]*n
result = 0

def dfs (v, depth):
    visited[v] = True
    global result
    # ABCDE의 친구 관계가 있다면
    if depth >= 4:
        result = 1
    else:
        # v 시작점부터 깊이를 탐색
        for i in friends[v]:
            if not visited[i]:
                dfs(i,depth+1)
        # 만약 마지막 정점이라면 False로 초기화
        visited[v] = False

for i in range(n):
    dfs(i, 0)
    if result:
        break

print(result)

File: Baek_joon_java/src/탐색/ABCDE.java
package 탐색;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class ABCDE {

    static ArrayList<Integer>[] arr;
    static boolean[] visited;
    static boolean isend;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        arr = new ArrayList[N];
        visited = new boolean[N];

        for (int i = 0; i < N; i++) {
            arr[i] = new ArrayList<>();
        }

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            arr[a].add(b);
            arr[b].add(a);
        }

        for (int i = 0; i < N; i++) {
            dfs(i, 1);

            if (isend) {
                break;
            }
        }
        System.out.println(isend ? 1 : 0);
    }

    private static void dfs(int friend, int depth) {
        if (depth == 5 || isend) {
            isend = true;
            return;
        }

        visited[friend] = true;
        for (int f : arr[friend]) {
            if (!visited[f]) {
                dfs(f, depth + 1);
            }
        }
        visited[friend] = false;
    }
}
